Implement more general initial guess mechanism

pbrehmer opened the pull request:
Enable implementation of an initial guess function in `Greedy`. It dispatches on the `info` tuple and on the arguments that were passed to the `assemble` call. This is important e.g. when using quantum number conserving DMRG, where specific initial states (states with defined QNs) are necessary. On this occasion, the `info` tuple will also be improved.

- closes #31 
- closes #38

pbrehmer commented:
@mfherbst This should be relatively ready to merge (from my side). Any thoughts so far?

mfherbst commented on src/greedy.jl:
```diff
@@ -31,34 +31,57 @@ Greedy reduced basis assembling strategy.
-    exit_checks::Bool  = true
+    tol::Float64      = 1e-3
+    n_truth_max::Int  = 64
+    Ψ_init::Function  = rb_guess
```
Putting this as `::Function` here is pointless as `::Function` is not a concrete type. I'd leave the type off.

pbrehmer replied:
Yes agreed, that would make more sense.

mfherbst commented on src/greedy.jl:
```diff
@@ -133,10 +154,12 @@ function assemble(info::NamedTuple, H::AffineDecomposition, grid, greedy::Greedy
         h_cache = HamiltonianCache(info.h_cache, ext.basis)
 
-        # Update iteration state info
+        # Update iteration state info (use Base.structdiff to remove keys from NamedTuple)
```
Hmm. Not sure ... how widespread is the use of this (non-exported) function. Can we rely on the interface to be stable?

pbrehmer replied:
Apparently this is used in a few packages and there are plans to export it, see https://github.com/JuliaLang/julia/issues/34772. So I would expect that this interface should stay stable, maybe up to a name change.

mfherbst commented on src/greedy.jl:
```diff
@@ -106,12 +129,10 @@ function assemble(info::NamedTuple, H::AffineDecomposition, grid, greedy::Greedy
-        end
-        truth = solve(H, μ_next, Ψ₀, solver_truth)
+        Ψ_init = greedy.Ψ_init((; info..., err_grid, λ_grid, err_max, μ=μ_next),
+                               (; H, grid, greedy, solver_truth, compressalg,
+                                solver_online))
```
Why does this need to be explicitly created? This looks a bit odd.

pbrehmer replied:
I wanted the initial guess function to have access to all `assemble(...)` arguments to keep the interface general, e.g. the RB initial guesses need to have access to the `solver_online` variable. The initial guess functions now have the signature `guess(info, offline_args)` which makes picking out specific arguments easy. I could also directly pass all offline arguments and then have `guess(info, offline_args...)` —  do you think this would make more sense?

mfherbst replied:
Hmm. Not sure I was more thinking it might be nicer to have all quantities in this named tuple be assigned to a variable rather than always constructing it in place. But if it's only used once let's keep it.

mfherbst commented on src/lobpcg.jl:
```diff
@@ -64,25 +64,21 @@ function FullDiagonalization(lobpcg::LOBPCG)
+function solve(H::AffineDecomposition, μ, Ψ₀::Matrix, lobpcg::LOBPCG)
+    !(size(Ψ₀, 1) == size(H, 1)) && error("Ψ₀ and H dimensions don't match")
+    if size(Ψ₀, 2) < lobpcg.n_target + lobpcg.n_ep_extra
+        Ψ₀_extra = randn(ComplexF64, size(Ψ₀, 1),
+                            lobpcg.n_target + lobpcg.n_ep_extra - size(Ψ₀, 2))
```
idention.

pbrehmer commented:
I now removed the `Function` type annotation and fixed the identation. I also happened to find an edge case in the `HamiltonianCache` constructor when I was experimenting today, which is now correctly handled.
